[pt2] add symbolic shape support for decompose mm and expose max_block to user config

jackiexu1992 opened the pull request:
Summary:
1) As described in https://fb.workplace.com/groups/1075192433118967/permalink/1381918665779674/
As a follow up, we can increase max_block["y"] to sovle the issue
2) add symbolic shape support for decompose mm pass. I did not find a good way to compare symint with int. So when there is a symbolic shape, i would assume it is a "large" dim.

Test Plan:
Without change block: aps-pt2-7c23cea900

increase y_block: aps-pt2_dynamic_shape-25a027423c

Differential Revision: D54525453




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

aakhundov commented:
> I did not find a good way to compare symint with int.

Maybe use [`V.graph.sizevars.size_hint`](https://github.com/pytorch/pytorch/blob/18d574a07a0af3057e7122beb2b2b7c3f3041fa1/torch/_inductor/sizevars.py#L417) to retrieve the corresponding concrete size from the example inputs behind the symbolic dim? Can also use a `fallback` if the symbolic dim is unbacked.

jackiexu77 committed:
[pt2] add symbolic shape support for decompose mm (#121440)

Summary:

 add symbolic shape support for decompose mm pass.
e.g. for coffee and first models, it can not decompose due to symbolic shape
```
Find mm mat1.meta['val'].shape = torch.Size([(((s0*((s1//5)))//128))*(((s0*((s1//5)))//(((s0*((s1//5)))//128)))), 2]), mat2.meta['val'].shape = torch.Size([2, 5])
```

Add some analysis for addmm decomposition. (result is plotted with num from D54669732) I plot the result for Mx2x2. When M is too small, decomposition is even slower. Hence, we have to check the longest dim.
 {F1466106449}

Test Plan:
Without change block: aps-pt2-7c23cea900

increase y_block: aps-pt2_dynamic_shape-25a027423c

After changes: torch.compile can run and the large linear can be decomposed
 {F1465916348}

baseline: without compile lce and decomposition
aps-baseline-c24a9ad0ae


Test on 60x aps first model: 
```
[trainer33|1]:[rank33]:I0315 21:47:04.952771 139825583611904 torch/fx/experimental/symbolic_shapes.py:3946] [8/18] eval 128*s0 >= 10240 [guard added] (_inductor/fx_passes/decompose_mem_bound_mm.py:55 in should_decompose_mm)
```
QPS 444K vs 480K 10% qps gain
 {F1469637315} 
aps-light_weight_1L_pt2_custom_decom-3be32e8eca
 {F1469593652}

Differential Revision: D54525453

huydhn commented:
@pytorchbot merge -i
